Vladimir-ReactJS-w3

intruston opened the pull request.

intruston commented:
Here's the link for deployed app https://luxury-toffee-40f2d8.netlify.app/

crevulus commented on week3/project/ecommerce/src/App.js:
```diff
@@ -0,0 +1,86 @@
+import React, { useState, useEffect } from "react";
```
Your `/src` folder is becoming quite large and hard to manage - do you think now would be a good time to start using subdirectories to store your files? `/hooks`, `/components`, etc.

intruston replied:
Did in my project, but left files as they were in this homework :)

crevulus commented on week3/project/ecommerce/src/App.js:
```diff
@@ -0,0 +1,86 @@
+            path="/"
+            element={
+              <>
+                <h1>Products</h1>
```
For the sake of human readability, it's best to pass a component name here (like you have done for products and favourites) rather than constructing a component inline. 🟠

intruston replied:
Done!

crevulus commented on week3/project/ecommerce/src/App.js:
```diff
@@ -0,0 +1,86 @@
+function App() {
+  const [categories, setCategories] = useState([]);
+  const [selectedCategory, setSelectedCategory] = useState(null);
+  const [favourites, setFavourites] = useState([]);
```
These values should be declared and stored in your context provider, not in `App`. Otherwise it will quickly become confusing where different values are stored. 🔴

intruston replied:
Agree, but every time I tried to move it my App failed.
But I fixed it and moved functions in CategoryAll (and pass the selectedCategory as props to ProductList) and in FavouritesContext

crevulus commented on week3/project/ecommerce/src/App.js:
```diff
@@ -0,0 +1,86 @@
+    <Router>
+      <div className="nav">
+        <Link to={"/"} className="nav-link">
+          {`Products`}
```
Why does this need to be in Javascript? It's jut a string. `{}` creates javascript logic inside your JSX. 🟢

intruston replied:
I forgot about it when I was troubleshooting some issues.

crevulus commented on week3/project/ecommerce/src/App.js:
```diff
@@ -0,0 +1,86 @@
+    }
+  };
+
+  const handleClick = (productId) => {
```
You should also move your adding/removing favourites logic to your context provider. 🔴

crevulus commented on week3/project/ecommerce/src/FavouriteList.js:
```diff
@@ -0,0 +1,64 @@
+            `https://fakestoreapi.com/products/${id}`
+          );
+          const product = await response.json();
+          return product;
```
What happens if there's an error? You should handle that as well. 🔴

intruston replied:
Add try-catch block

crevulus commented on week3/project/ecommerce/src/useFetch.js:
```diff
@@ -0,0 +1,31 @@
+import { useState, useEffect } from "react";
+
+function useFetch(url) {
```
Great hook, well done!

crevulus commented on week3/project/ecommerce/src/ProductList.js:
```diff
@@ -0,0 +1,52 @@
+    url = "https://fakestoreapi.com/products/category/" + selectedCategory;
+  }
+
+  const { data, isLoading, error } = useFetch(url, [selectedCategory]);
```
Here you're passing two arguments, but your fetch only takes one. `selectedCategory` does nothing here. 🔴

intruston replied:
Agree, probably thought about useEffect, when I was writing this line :)
Fixed.

crevulus reviewed (commented):
Hi Vladimir,

Great application, once again. You clearly have a strong understanding of react!

You should develop your context a little bit more. Currently it's just storing values, when relly you should also include functionality like setting and unsetting favourites in the context provider. By keeping everyting in `App`, your react application will quickly become unmanageable as it grows. So do be careful of that.

crevulus reviewed (commented):
Hi Vladimir,

Great application, once again. You clearly have a strong understanding of react!

You should develop your context a little bit more. Currently it's just storing values, when relly you should also include functionality like setting and unsetting favourites in the context provider. By keeping everyting in `App`, your react application will quickly become unmanageable as it grows. So do be careful of that.

intruston commented on week3/project/ecommerce/src/useFetch.js:
```diff
@@ -0,0 +1,31 @@
+        if (!response.ok) {
+          throw new Error("Failed to fetch");
+        }
+        const json = await response.json();
```
Thanks! Updated!

intruston commented:
@crevulus , thanks for your review. Most of the issues are about overwhelmed App.js, so I fixed it and moved functions in CategoryAll (and pass the selectedCategory as props to ProductList) and in FavouritesContext.
Now I think, everything is fine!
